### ISSUE TITLE
xgboost vs xgbtree

### ISSUE THREAD

**Issue by philipus** [topepo/caret#297]
I have got serious problems using xgbtree with caret especially with additional options like min_child_weight

when min_chil_weight is bigger than 50 i have got a roc of 0.5 for the complete parater (eta, nrounds, max_depth) room

in comparison with xgboost I got good (auc differs from random) results with the same parameter settings

could it be because auf the roc metric which is not implemented in xgboost????

here are my parameter setting for xbgtree:

xgbGrid <- expand.grid(nrounds = c(60),
                       max_depth = c(17),
                       eta = c(0.01))

xgbfitContr <- trainControl(method = "cv", number = 3, returnResamp = "all",
                            classProbs = TRUE,
                            summaryFunction = twoClassSummary
                            )

set.seed(849)

xgbFit <- train(target ~ ., 
                data = training[,c("target",feature.names)],
                method = "xgbTree",
                trControl = xgbfitContr,
                tuneGrid = xgbGrid,
                metric = "ROC",
                subsample = 0.2,    # part of data instances to grow tree 
                colsample_bytree = 0.7,  # subsample ratio of columns when constructing each tree 
                min_child_weight = 500,
                base_score = 0.03304549,
                missing = NA
)

and here is the parameter setting in xgboost

param <- list("objective" = "binary:logistic",    # multiclass classification 
              "eval_metric" = "auc",    # evaluation metric 
              "max_depth" = 17,    # maximum depth of tree 
              "eta" = 0.01,    # step size shrinkage 
              "subsample" = 0.2,    # part of data instances to grow tree 
              "colsample_bytree" = 0.7,  # subsample ratio of columns when constructing each tree 
              "min_child_weight" = 500

)

bst <- xgboost(param=param, 
               data = data.matrix(data[,feature.names]), 
               label=data$p_aktiv, 
               nrounds=60,
               base_score = 0.03304549,
               missing=NA
)

post scriptum: p_aktiv is 0 and 1. for caret I made the following mutation befor

data %<>% do(mutate(., target = factor(ifelse(p_aktiv==1,"one","zero"))))

so target is factor named by one and zero because with names in the target variable like "0" and "1" it doesnt work...

**Comment by topepo** [comment]
Can you describe the data more? It could be anything related to the data. For example, is there a large class imbalance? This might yield good accuracy (by calling everything a non-event) but be bad from the standpoint of the ROC curve.

**Comment by philipus** [comment]
yeah

extremely  imbalanced. response rate of 3.5%. means 3.5% with 1 and 96.5% with 0

1 mio sets with 500 columns. thats why I need to set min_child_weight >500 otherwise I would get overfitting

which metric is applied in caret?

can one individualie it?

**Comment by topepo** [comment]
`train` uses whatever metric you pick to measure performance and pick the best model. That might not be the same loss function/criterion that the underlying model is optimizing though. 

If you want to optimize the tuning parameters on the area under the ROC curve, look at the `twoClassSummary` function (examples are [here](http://topepo.github.io/caret/training.html#metrics)).

**Comment by philipus** [comment]
yeah I had a look at it. can you control the metric (eval on valid) by summaryFunction  ???

something doesnt work. I will check ist and tell you tomorrow

**Comment by topepo** [comment]
You can control the metric that `train` uses to measure performance and optimize the model using the `summaryFunction` argument in `trainControl`. As for the underlying model, it depends on what that function allows you to do.

**Comment by philipus** [comment]
i think i got the problem here -> most of my data rows have missing values in their columns

I use "missing = NA" in xgboost. so no imputation -> the same with caret

i noticed when I predict the target using the caret  fit with type="prob"

i just get very view prediction... i didnt find the part where the evaluation takes place in the caret code

also tried to use : predict(xgbFit, training[,feature.names], type="prob", na.action = na.pass)

didnt help

can you give me a hint where I find the evaluation part in caret

**Comment by topepo** [comment]
We really need a reproducible example to help.

**Comment by philipus** [comment]
yeah I can imagine! hm... i will make a probe

i guess here is the problem -> newdata <- predict(preProc, newdata) (see below)

anyway I'll make the probe for you


##################

probFunction <- function(method, modelFit, newdata, preProc = NULL, param = NULL)
{
  if(!is.null(preProc)) newdata <- predict(preProc, newdata)

  obsLevels <- levels(modelFit)
  
  classProb <- method$prob(modelFit = modelFit, 
                           newdata = newdata, 
                           submodels = param)  
  if(!is.data.frame(classProb) & is.null(param))
  {
    classProb <- as.data.frame(classProb)
    if(!is.null(obsLevels)) classprob <- classProb[, obsLevels]
  }
  classProb
}

**Comment by philipus** [comment]
drop box ok for you or how do you wonna get the data from me
